### PR TITLE
Return of the build script via Preferences

### PR DESCRIPTION
**Pull request by mkitti**
This restores the build script to facilitate downloading the GR binary. It is currently imported via GRPreferences. The main purpose of restoring the build script is as a fallback when the BinaryBuilder binary does not work. For example, Wayland support is currently lacking in the BinaryBuilder QT builds.

Currently, this script can be invoked via

```
using GR
GR.GRPreferences.use_upstream_binary()
```

I'll probably change the function name to make it clearer that this downloads a binary.

cc: @t-bltg

### CODE REVIEW

**Comment by mkitti** [comment]
Wayland issues xref: https://github.com/sciapp/gr/issues/141#

**Comment by mkitti** [comment]
cc: @sjkelly , it turns out the alternate binaries are still useful for Wayland support.

**Comment by t-bltg** [comment]
I think that the build script can be simplified, instead of just copy-paste what was removed in https://github.com/jheinen/GR.jl/pull/465.

**Comment by sjkelly** [comment]
This sounds reasonable to me. I appreciate everyone filling in the gaps in my artifacts PR. I assume this doesn't break precompilation on 1.8+?

**Comment by mkitti** [comment]
> I think that the build script can be simplified, instead of just copy-paste what was removed in #465.

Yes, I'm working on it. I just wanted to make to develop a proof of concept to see if the old script still works before I starting cutting it down.

> This sounds reasonable to me. I appreciate everyone filling in the gaps in my artifacts PR. I assume this doesn't break precompilation on 1.8+?

I have not modified the `__init__` process. We're entirely using the Preferences.jl mechanism still. Basically what I'm restoring is the ability to automatically download the correct upstream GR tarball.

To decide:
1. Name of the download function. `use_deps_binary` does not seem descriptive.
2. Where to put the downloaded tarball. I'm not sure if the deps folder is the right place.
3. We might be able to use the artifacts system still.... just specifying these as custom artifacts.

**Comment by t-bltg** [comment]
> I assume this doesn't break precompilation on 1.8+

Xref https://github.com/timholy/SnoopCompile.jl/issues/295, this would have to be tested against `Plots` precompilation mechanism using `SnoopPrecompile`.

**Comment by jheinen** [comment]
Yes, this would simplify the installation of the GR tarball for many users. However, the runtime problem under Linux still exists. I could only achieve correct results with the `LD_PRELOAD` environment variable (see https://github.com/sciapp/gr/issues/141#issuecomment-1298664206)

**Comment by mkitti** [comment]
> However, the runtime problem under Linux still exists. I could only achieve correct results with the `LD_PRELOAD` environment variable

Addressing https://github.com/jheinen/GR.jl/issues/484 may resolve the issue. Basically if we use `dont_dlopen=true`, then the GR_jll binaries will never be loaded if Preferences points to the other binaries.

**Comment by jheinen** [comment]
@mkitti : Thanks for your PR.

Would it be possible to "control" the behaviour from outside Julia, e.g.
`GRDIR=/usr/local/gr Julia ...`
without restarting Julia (when preferences have been changed).

The background of my question is: We use Julia in a read-only network environment with pre-installed packages (which are maintained daily) for all users. Possibly a strange requirement, but the process has proven itself.

**Comment by mkitti** [comment]
> Would it be possible to "control" the behaviour from outside Julia, e.g. `GRDIR=/usr/local/gr julia ...` without restarting Julia (when preferences have been changed).

Yes. Essentially you could preconfigure by providing an Overrides.toml in the depot and a LocalPreferences.toml in the project.

## Depot based Overrides.toml

The first mechanism is older and allows for overriding the binary locations via a depot based Overrides.toml [1].

```julia
julia> using GR
[ Info: Precompiling GR [28b8d3ca-fb5f-59d9-8090-bfdbd6d07a71]

julia> GR.GRPreferences.use_upstream_binary(; force = true, override = :depot)
┌ Info: Downloading pre-compiled GR 0.69.1 Ubuntu binary
└   file = "/tmp/jl_rbiTw2/gr-0.69.1-Ubuntu-x86_64.tar.gz"
┌ Info: grdir
└   grdir = "/home/mkitti/.julia/dev/GR/deps/gr"
[ Info: Please restart Julia to change the GR binary configuration.

```

What that does is create an Overrides.toml in the Julia depot:

```
$ cat ~/.julia/artifacts/Overrides.toml
[d2c73de3-f751-5644-a686-071e5b155ba9]
GR = "/home/mkitti/.julia/dev/GR/deps/gr"
```

Thus, one option that you have is to simply create that Overrides.toml in the depot. The UUID there is that of GR_jll:
https://github.com/JuliaBinaryWrappers/GR_jll.jl/blob/fc933dae4a54c49707193e2dcc7df1e5ef201cab/Project.toml#L2

```julia
julia> GR_jll.libGR
"/home/mkitti/.julia/dev/GR/deps/gr/lib/libGR.so"

julia> GR_jll.libGR3
"/home/mkitti/.julia/dev/GR/deps/gr/lib/libGR3.so"

julia> GR_jll.libGRM
"/home/mkitti/.julia/dev/GR/deps/gr/lib/libGRM.so"

julia> GR_jll.gksqt_path
"/home/mkitti/.julia/dev/GR/deps/gr/bin/gksqt"

julia> DEPOT_PATH
3-element Vector{String}:
 "/home/mkitti/.julia"
 "/home/mkitti/src/julia-1.8.1/local/share/julia"
 "/home/mkitti/src/julia-1.8.1/share/julia"
```

Note that above not only have we instructed GR to load the alternate binaries, but we've now also redirected GR_jll to alternate binaries. This should comprehensively take care of the binary issue. You should not need `LD_PRELOAD` because GR_jll actually is loading the same binaries that we want. Yes, this is currently a bit redundant.

Recall also that depots stack on the system depots. You could also locate the Overrides.toml in
```
/home/mkitti/src/julia-1.8.1/local/share/julia/artifacts/Overrides.toml
/home/mkitti/src/julia-1.8.1/share/julia/artifacts/Overrides.toml
```

It also still makes the modification to the LocalPreferences.toml stored in the same directory as the Project.toml to memorize this configuration setting:
```
$ cat LocalPreferences.toml 
[GR]
binary = "system"
grdir = "/home/mkitti/.julia/dev/GR/deps/gr"
```

## Project based LocalPreferences.toml configuration

Another option involves only using LocalPreferences.toml, stored in the project [2]. Here I've installed GR into `/home/mkitti/gr`.

```
julia> using GR

julia> GR.GRPreferences.use_upstream_binary("/home/mkitti"; force = true, override = :project)
┌ Info: Downloading pre-compiled GR 0.69.1 Ubuntu binary
└   file = "/tmp/jl_4IzraC/gr-0.69.1-Ubuntu-x86_64.tar.gz"
┌ Info: grdir
└   grdir = "/home/mkitti/gr"
[ Info: Please restart Julia to change the GR binary configuration.
```

Here the paths to the individual binaries is stored in the LocalPreferences.toml.
```
$ cat LocalPreferences.toml 
[GR]
binary = "system"
grdir = "/home/mkitti/gr"

[GR_jll]
gksqt_path = "/home/mkitti/gr/bin/gksqt"
libGKS_path = "/home/mkitti/gr/lib/libGKS"
libGR3_path = "/home/mkitti/gr/lib/libGR3"
libGRM_path = "/home/mkitti/gr/lib/libGRM"
libGR_path = "/home/mkitti/gr/lib/libGR"
```

## Summary

Yes, you can control which binaries are loaded by defining either `JULIA_DEPOT_PATH/artifacts/Overrides.toml` or `/path/to/project/LocalPreferences.toml` as above. Writing the TOML files can be done completely outside Julia. Before this PR, the LocalPreferences.toml only redirected GR to use different binaries. With this PR, GR_jll is now also redirected to use alternate binaries. This in part addresses the `LD_PRELOAD` issue.

## References

[1] Overrides.toml in the Pkg.jl documentation. https://pkgdocs.julialang.org/v1/artifacts/#Overriding-artifact-locations
[2] (LocalPreferences.toml) in the BinaryBuilder.jl documentation. https://docs.binarybuilder.org/stable/jll/#Non-dev'ed-JLL-packages

**Comment by mkitti** [comment]
> Would it be possible to "control" the behaviour from outside Julia, e.g.
> `GRDIR=/usr/local/gr julia ...`

To clarify, the `GRDIR` environment variable override will also continue to function, but we need https://github.com/JuliaPackaging/Yggdrasil/pull/5784 to avoid the need for `LD_PRELOAD`.

**Comment by jheinen** [comment]
@mkitti : Thanks for the detailed explanations. If I can merge, just give a quick shout.

File: src/downloader.jl [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+"""
+    get_default_install_dir()
+
+Return the default install directory where we have write permissions.
+Currently, this is the deps directory of the GR package:
+`joinpath(pathof(GR), "..", "deps")`
+"""
+function get_default_install_dir()
+    return abspath(joinpath(@__DIR__, "..", "deps"))
+end
```
Comment (t-bltg) [inline src/downloader.jl:260]:
```suggestion
get_default_install_dir() =
    abspath(joinpath(@__DIR__, "..", "deps"))
```

Let's be concise.

File: src/downloader.jl [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    @info("Downloading pre-compiled GR $version $os binary", file)
+    # Download versioned tarballs from Github
+    if version != "latest"
+        ok = try_download(
+            "https://github.com/sciapp/gr/releases/download/v$version/$tarball",
+            file,
+        )
+    else
+        ok = false
+    end
```
Comment (t-bltg) [inline src/downloader.jl:280]:
```suggestion
    ok = if version != "latest"
        try_download(
            "https://github.com/sciapp/gr/releases/download/v$version/$tarball",
            file,
        )
    else
        false
    end
```

File: src/downloader.jl [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+    # Download latest tarball from gr-framework.org
+    if !ok
+        if !try_download("https://gr-framework.org/downloads/$tarball", file)
+            @warn("Using insecure connection")
```
Comment (t-bltg) [inline src/downloader.jl:285]:
```suggestion
            @warn "Using insecure connection"
```

File: src/downloader.jl [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+    # Ensure the working directory exists
+    mkpath(install_dir)
+
+    if force
+        # Download regardless if an existing installation exists
+        grdir = nothing
+    else
+        # Check for an existing installation
+        grdir = get_grdir()
+    end
```
Comment (t-bltg) [inline src/downloader.jl:318]:
```suggestion
    grdir = if force
        # Download regardless if an existing installation exists
        nothing
    else
        # Check for an existing installation
        get_grdir()
    end
```

File: src/preferences.jl [hunk 5]
```diff
@@ -53,19 +69,162 @@ module GRPreferences
...
+            if err isa ArgumentError
+                throw(ArgumentError("Could not set GR system binary preference. Consider using the `force = true` keyword argument."))
+            end
+        end
+        if override isa Symbol
+            override = (override,)
+        end
+        if :depot in override
+            override_depot(grdir)
+        end
```
Comment (t-bltg) [inline src/preferences.jl:102]:
```suggestion
        :depot in override && override_depot(grdir)
```

File: src/preferences.jl [hunk 6]
```diff
@@ -53,19 +69,162 @@ module GRPreferences
...
+        if :project in override
+            override_project(grdir; force)
+        end
```
Comment (t-bltg) [inline src/preferences.jl:105]:
```suggestion
        :project in override && override_project(grdir; force)
```

File: src/downloader.jl [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+    try
+        gksqt = joinpath(grdir, "bin", "gksqt")
+        res = read(`ldd $gksqt`, String)
+        if occursin("not found", res)
+            @warn(
+                "Missing dependencies for GKS QtTerm. Did you install the Qt5 run-time?"
+            )
+            @warn(
+                "Please refer to https://gr-framework.org/julia.html for further information."
+            )
```
Comment (t-bltg) [inline src/downloader.jl:208]:
```suggestion
            @warn """
            Missing dependencies for GKS QtTerm. Did you install the Qt5 run-time ?
            Please refer to https://gr-framework.org/julia.html for further information.
            """
```

File: src/downloader.jl [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+    )
+    try
+        @static if VERSION >= v"1.4.0-DEV.265"
+            have_qml = haskey(
+                Pkg.dependencies(),
+                UUID("2db162a6-7e43-52c3-8d84-290c1c42d82a"),
+            )
+        else
+            have_qml = haskey(Pkg.API.installed(), "QML")
+        end
```
Comment (t-bltg) [inline src/downloader.jl:233]:
```suggestion
        have_qml = @static if VERSION >= v"1.4.0-DEV.265"
            haskey(
                Pkg.dependencies(),
                UUID("2db162a6-7e43-52c3-8d84-290c1c42d82a"),
            )
        else
            haskey(Pkg.API.installed(), "QML")
        end
```

File: src/downloader.jl [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+"""
+function get_version()
+    _version = version
+    if "GRDIR" in keys(ENV)
+        if isempty(ENV["GRDIR"])
+            _version = "latest"
+        end
+    end
+    return _version
+end
```
[the remark this answers is not in the document]
Comment (mkitti) [reply src/downloader.jl:72]: Any thoughts about this @jheinen . Do we need the version to be `latest` if the `GRDIR` environment variable is set?

[the remark this answers is not in the document]
Comment (jheinen) [reply src/downloader.jl:72]: No, for development purposes, there are probably better ways to test the latest run-time. End users don't need this.

File: src/downloader.jl [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+subdirectory.
+
+The function will return a `String` representing one of the paths above
+or `nothing` if a GR install is not located in any of those locations.
+"""
+function get_grdir()
+    if "GRDIR" in keys(ENV)
+        grdir = ENV["GRDIR"]
+        have_dir = length(grdir) > 0
+        if have_dir
```
Comment (t-bltg) [inline src/downloader.jl:38]:
```suggestion
        if (have_dir = length(grdir) > 0)
```

File: src/downloader.jl [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+            grdir = joinpath(d, "gr")
+            if isdir(joinpath(grdir, "fonts"))
+                have_dir = true
+                break
+            end
+        end
+    end
+    if have_dir
+        @info("Found existing GR run-time in $grdir")
+    end
```
Comment (t-bltg) [inline src/downloader.jl:53]:
```suggestion
    have_dir && @info "Found existing GR run-time in $grdir"
```

File: src/downloader.jl [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+    get_os_and_arch()
+
+Figure out which specific operating system this, including the specific Linux distribution.
+"""
+function get_os_and_arch()
+    if Sys.iswindows()
+        os = "Windows"
+    else
+        os = string(Sys.KERNEL)
+    end
```
Comment (t-bltg) [inline src/downloader.jl:103]:
```suggestion
    os = Sys.iswindows() ? "Windows" : string(Sys.KERNEL)
```

File: src/downloader.jl [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+
+    # Download latest tarball from gr-framework.org
+    if !ok
+        if !try_download("https://gr-framework.org/downloads/$tarball", file)
+            @warn "Using insecure connection"
+            if !try_download("http://gr-framework.org/downloads/$tarball", file)
+                error("Cannot download GR run-time")
+            end
+        end
+    end
```
Comment (t-bltg) [inline src/downloader.jl:287]:
```suggestion
    if !ok && !try_download("https://gr-framework.org/downloads/$tarball", file)
        @warn "Using insecure connection"
        try_download("http://gr-framework.org/downloads/$tarball", file) ||
            error("Cannot download GR run-time")
    end
```

File: src/downloader.jl [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+                Tar.extract(`$(p7zip_jll.p7zip()) x $file -so`, extract_dir)
+                mv(joinpath(extract_dir, "gr"), destination_dir)
+            end
+            rm(file)
+        end
+
+        # Address Mac specific framework and rpath issues
+        if Sys.isapple()
+            apple_install(destination_dir)
+        end
```
Comment (t-bltg) [inline src/downloader.jl:345]:
```suggestion
        Sys.isapple() && apple_install(destination_dir)
```

File: src/downloader.jl [hunk 15]
```diff
@@ -0,0 +1,360 @@
...
+
+        grdir = destination_dir
+    end # if isnothing(grdir)
+
+    # Check dependencies when using Linux or FreeBSD
+    if Sys.islinux() || Sys.isfreebsd()
+        check_dependencies(grdir)
+    end
```
Comment (t-bltg) [inline src/downloader.jl:353]:
```suggestion
    (Sys.islinux() || Sys.isfreebsd()) && check_dependencies(grdir)
```

File: src/downloader.jl [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+subdirectory.
+
+The function will return a `String` representing one of the paths above
+or `nothing` if a GR install is not located in any of those locations.
+"""
+function get_grdir()
+    if "GRDIR" in keys(ENV)
+        grdir = ENV["GRDIR"]
+        have_dir = length(grdir) > 0
+        if have_dir
```
Comment (mkitti) [inline src/downloader.jl:38]: How about this?

```suggestion
        if (have_dir = !isempty(grdir))
```

**Comment by mkitti** [comment]
The functionality is where I would like it.

However, I think we should still do the following:
1. Add tests
2. Add documentation

What is the best way to update the Sphinx documentation at gr-framework.org? Otherwise, I'll just update the README.

Perhaps we should implement Documenter.jl based documentation?

**Comment by jheinen** [comment]
> What is the best way to update the Sphinx documentation at gr-framework.org? Otherwise, I'll just update the README.

To get ahead quickly, the README file should be updated. The official GR documentation is unfortunately only in a private repo.

> Perhaps we should implement Documenter.jl based documentation?

Would probably make sense in the medium term. I would have to find out how to integrate this sensibly.

**Comment by mkitti** [comment]
> Would probably make sense in the medium term. I would have to find out how to integrate this sensibly.

I can help open another pull request on this after we merge this.

**Comment by mkitti** [comment]
There definitely seems to be some kind of race condition going on with `Volume plot`

**Comment by mkitti** [comment]
@jheinen Let's merge this as soon as possible since this fixes the plugin breakage as described in #486.

**Pull request merged**